Add utility function to create vector of pairs from two input parameters

parikshitbajpai opened the pull request:
- Adds a helper function `getParamPairs` to `MooseObject` that takes two parameters of type std::vector<T1> and std::vector<T2>
```
std::vector<pair<T1,T2> getParamPairs<T1,T2>(const std::string & param1, const std::string & param2);
```
- Performs error checking for vector size, splices them into a single vector of pairs and returns the vector.

closes #18456

Commit:
Add utility function to create vector of pairs from two input parameters

- Add helper function `getParamPairs`
- Creates vector of pairs instead of two vectors for a list of 
corresponding pairs

Fixes #18456

Commit:
Modify paramater pair error message to show location of parameter

dschwen commented on framework/include/base/MooseObject.h:
```diff
@@ -87,6 +87,16 @@ class MooseObject : public ConsoleStreamInterface, public libMesh::ParallelObjec
+   * @return Vector of pairs of first and second parameters
+   */
+  template <typename T1, typename T2>
+  std::vector<std::pair<T1, T2>> getParamPair(const std::string & param1,
```
I would call it `getParamPairs`

dschwen commented on framework/include/base/MooseObject.h:
```diff
@@ -255,6 +265,36 @@ MooseObject::getParam(const std::string & name) const
+std::vector<std::pair<T1, T2>>
+MooseObject::getParamPair(const std::string & param1, const std::string & param2) const
+{
+  std::vector<std::pair<T1, T2>> parameter_pair;
+  std::string msg;
+
+  if (isParamValid(param1) && isParamValid(param2))
+  {
+    auto v1 = getParam<std::vector<T1>>(param1);
+    auto v2 = getParam<std::vector<T2>>(param2);
+
+    if (v1.size() == v2.size())
+      for (std::size_t i = 0; i < v1.size(); ++i)
+        parameter_pair.emplace_back(std::make_pair(v1[i], v2[i]));
+    else
+      msg = "Sizes of vector parameters " + paramErrorPrefix(_pars, param1) + " and " +
+            paramErrorPrefix(_pars, param2) + " are unequal \n";
+  }
+  else if (!isParamValid(param1))
+    msg = paramErrorMsg(param1, "Invalid parameter");
+  else if (!isParamValid(param2))
+    msg = paramErrorMsg(param2, "Invalid parameter");
+
+  if (!msg.empty())
+    callMooseErrorRaw(msg, &_app);
+  else
+    return parameter_pair;
```
```suggestion
  if (!isParamValid(param1))
    paramError(param1, "Invalid parameter");
  if (!isParamValid(param2))
    paramError(param2, "Invalid parameter");

  auto v1 = getParam<std::vector<T1>>(param1);
  auto v2 = getParam<std::vector<T2>>(param2);

  if (v1.size() != v2.size())
  {
    callMooseErrorRaw("Sizes of vector parameters " + paramErrorPrefix(_pars, param1) + " and " +
            paramErrorPrefix(_pars, param2) + " are unequal \n", &_app);
  }
  
  std::vector<std::pair<T1, T2>> parameter_pair;
  for (std::size_t i = 0; i < v1.size(); ++i)
    parameter_pair.emplace_back(std::make_pair(v1[i], v2[i]));
  return parameter_pair;
```

parikshitbajpai committed:
Restructure error message for compactness

Co-authored-by: Daniel Schwen <daniel@schwen.de>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/785816/) on 85767a3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 58219425a3180a3429cc37d840a201fa3acf58f6`

Commit:
Rename `getParamPair` to `getparamPairs`

Commit:
Edit error messages in `getParamPairs`

- Remove `isParamValid()` checks and associated error messages
- Add vector sizes to unequal vectors error

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/785887/) on a7dd4e6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18462/site)

This comment will be updated on new commits.
